File: packages/java/anip-crypto/src/main/java/dev/anip/crypto/JwsSigner.java
package dev.anip.crypto;

import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.JWSObject;
import com.nimbusds.jose.Payload;
import com.nimbusds.jose.crypto.ECDSASigner;
import com.nimbusds.jose.crypto.ECDSAVerifier;
import com.nimbusds.jose.util.Base64URL;

import java.text.ParseException;

/**
 * Handles detached JWS signing and verification for the X-ANIP-Signature header.
 */
public final class JwsSigner {

    private JwsSigner() {}

    /**
     * Creates a detached JWS signature for the given payload.
     * The result is a compact JWS with an empty payload: "header..signature".
     * Used for the X-ANIP-Signature header on manifest responses.
     *
     * @param km      the key manager (uses delegation key)
     * @param payload the payload bytes to sign
     * @return the detached JWS string in format "header..signature"
     */
    public static String signDetachedJws(KeyManager km, byte[] payload) throws JOSEException {
        JWSHeader header = new JWSHeader.Builder(JWSAlgorithm.ES256)
                .keyID(km.getDelegationKid())
                .build();

        JWSObject jwsObject = new JWSObject(header, new Payload(payload));
        ECDSASigner signer = new ECDSASigner(km.getDelegationECKey());
        jwsObject.sign(signer);

        // Serialize as compact, then remove the payload part to make it detached
        String compact = jwsObject.serialize();
        // compact format is: header.payload.signature
        // detached format is: header..signature
        String[] parts = compact.split("\\.", 3);
        return parts[0] + ".." + parts[2];
    }

    /**
     * Verifies a detached JWS signature against the given payload.
     *
     * @param km        the key manager (uses delegation public key)
     * @param payload   the original payload bytes
     * @param signature the detached JWS string ("header..signature")
     * @throws JOSEException if verification fails
     */
    public static void verifyDetachedJws(KeyManager km, byte[] payload, String signature)
            throws JOSEException, ParseException {

        // Split the detached JWS: "header..signature"
        String[] parts = signature.split("\\.", -1);
        if (parts.length != 3) {
            throw new JOSEException("Invalid detached JWS format: expected 3 parts");
        }
        if (!parts[1].isEmpty()) {
            throw new JOSEException("Invalid detached JWS: payload part should be empty");
        }

        // Reconstruct the full JWS by inserting the payload
        Base64URL payloadB64 = Base64URL.encode(payload);
        String fullJws = parts[0] + "." + payloadB64 + "." + parts[2];

        JWSObject jwsObject = JWSObject.parse(fullJws);
        ECDSAVerifier verifier = new ECDSAVerifier(km.getDelegationECKey().toECPublicKey());

        if (!jwsObject.verify(verifier)) {
            throw new JOSEException("Invalid signature");
        }
    }
}

File: packages/csharp/src/Anip.Crypto/JwsSigner.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Anip.Crypto;

/// <summary>
/// Creates detached JWS signatures for payload integrity verification.
/// Used for the X-ANIP-Signature header on manifest responses.
/// </summary>
public static class JwsSigner
{
    /// <summary>
    /// Creates a detached JWS signature for the given payload.
    /// The result is a compact JWS with an empty payload section: "header..signature".
    /// </summary>
    /// <param name="privateKey">EC P-256 private key for signing.</param>
    /// <param name="keyId">Key ID to include in the JWS header.</param>
    /// <param name="payload">The payload bytes to sign.</param>
    /// <returns>A detached JWS string in the format "header..signature".</returns>
    public static string SignDetached(ECDsa privateKey, string keyId, byte[] payload)
    {
        var header = new Dictionary<string, string>
        {
            ["alg"] = "ES256",
            ["kid"] = keyId
        };

        var headerJson = JsonSerializer.Serialize(header);
        var headerB64 = KeyManager.Base64UrlEncode(Encoding.UTF8.GetBytes(headerJson));
        var payloadB64 = KeyManager.Base64UrlEncode(payload);

        // The signing input includes the payload, but the output omits it.
        var signingInput = $"{headerB64}.{payloadB64}";
        var signingInputBytes = Encoding.UTF8.GetBytes(signingInput);

        var signatureBytes = privateKey.SignData(signingInputBytes, HashAlgorithmName.SHA256, DSASignatureFormat.IeeeP1363FixedFieldConcatenation);

        var signatureB64 = KeyManager.Base64UrlEncode(signatureBytes);

        // Detached: omit the payload part.
        return $"{headerB64}..{signatureB64}";
    }

    /// <summary>
    /// Verifies a detached JWS signature against the given payload.
    /// </summary>
    /// <param name="publicKey">EC P-256 public key for verification.</param>
    /// <param name="payload">The original payload bytes.</param>
    /// <param name="signature">The detached JWS string ("header..signature").</param>
    /// <returns>True if the signature is valid.</returns>
    public static bool VerifyDetached(ECDsa publicKey, byte[] payload, string signature)
    {
        var parts = signature.Split('.', 4);
        if (parts.Length != 3 || parts[1] != "")
        {
            return false;
        }

        var headerB64 = parts[0];
        var sigB64 = parts[2];

        // Decode and validate header.
        try
        {
            var headerJson = Encoding.UTF8.GetString(KeyManager.Base64UrlDecode(headerB64));
            var header = JsonSerializer.Deserialize<Dictionary<string, string>>(headerJson);
            if (header == null || !header.TryGetValue("alg", out var alg) || alg != "ES256")
            {
                return false;
            }
        }
        catch
        {
            return false;
        }

        // Reconstruct signing input with the payload.
        var payloadB64 = KeyManager.Base64UrlEncode(payload);
        var signingInput = $"{headerB64}.{payloadB64}";
        var signingInputBytes = Encoding.UTF8.GetBytes(signingInput);

        try
        {
            var signatureBytes = KeyManager.Base64UrlDecode(sigB64);
            return publicKey.VerifyData(signingInputBytes, signatureBytes, HashAlgorithmName.SHA256, DSASignatureFormat.IeeeP1363FixedFieldConcatenation);
        }
        catch
        {
            return false;
        }
    }
}
